### PR TITLE
Added recursive `pathPrefix` in contract options

### LINKED ISSUE #215: Prefixed URL
Hello guys, there is a way to use prefixed url or nested calls in the router, like this?

```
class App {
  public app: express.Application;

  constructor() {
    this.app = express();
    this.config();
    this.routerSetup();
    this.errorHandler();
  }

  private config() {
    // view engine setup
    this.app.set('views', path.join(__dirname, 'views'));
    this.app.set('view engine', 'jade');

    this.app.use(logger('dev'));
    this.app.use(express.json());
    this.app.use(express.urlencoded({ extended: false }));
    this.app.use(cookieParser());
    this.app.use(express.static(path.join(__dirname, 'public')));
  }

  private routerSetup() {
    this.app.use('/', indexRouter);
    this.app.use('/users', usersRouter);
    this.app.use(V1(this.app, '/v1'));
  }

```

```

const c = initContract();

export const postContract = c.router({
  createPost: {
    method: 'POST',
    path: '/posts',
    responses: {
      201: c.response<Post>(),
    },
    body: c.body<{title: string}>(),
    summary: 'Create a post',
  },
  getPost: {
    method: 'GET',
    path: `/posts/:id`,
    responses: {
      200: c.response<Post | null>(),
    },
    summary: 'Get a post by id',
  },
});

```

```
export function V1(app: Application, url: string): Router {
  return Router().use('/test', PostHandler.register());
}
```

```
    static register() {
        return Router()
            .use(
                '/groups/:groupId/post',
                GroupHandler.register(connection)
            )
            .use('/groups/:groupId/post/:id/comments', GroupCommentHandler.register())
            .use(
                '/groups/:groupId/post/:id',
                GroupHandler.register()
            )}
```

I want those urls in that file to be  like 'v1/test/groups/:groupId/post' without having to rewrite v1/test every time.
Because some times I cannot be sure that will be v1/test, since I'm creating a larger api with many modules.
Maybe make createExpressEndpoints reusable to make a chain of "router.use".

### PR DESCRIPTION
**Pull request by Jwcode-uk**
closes #215

Added changes to allow user to supply path prefix in options and have it added to path recursively.

### CODE REVIEW

File: libs/ts-rest/core/src/lib/dsl.spec.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /* eslint-disable @typescript-eslint/no-unused-vars */
-import { z } from 'zod';
+import { TypeOf, z } from 'zod';
```
Comment (oliverbutler) [inline libs/ts-rest/core/src/lib/dsl.spec.ts:2]: Can we remove this

File: libs/example-microservice/util-posts-api/src/lib/example-microservice-util-posts-api.ts [hunk 2]
```diff
@@ -19,35 +19,38 @@ export type Post = z.infer<typeof PostSchema>;
...
+          message: z.string(),
+        }),
+      },
+      description: 'Update post thumbnail',
     },
-    description: 'Update post thumbnail',
   },
-});
+  { urlPrefix: 'test' }
+);
```
Comment (oliverbutler) [inline libs/example-microservice/util-posts-api/src/lib/example-microservice-util-posts-api.ts:56]: Why did we change this?

**Comment by oliverbutler** [comment]
Damn - that was quick 🎉 Great to see this was q quick win

cat-speed

File: libs/example-microservice/util-posts-api/src/lib/example-microservice-util-posts-api.ts (hunk 2 above)
Comment (Jwcode-uk) [reply libs/example-microservice/util-posts-api/src/lib/example-microservice-util-posts-api.ts]: No Comment 👀

File: libs/ts-rest/core/src/lib/dsl.spec.ts (hunk 1 above)
Comment (Jwcode-uk) [reply libs/ts-rest/core/src/lib/dsl.spec.ts]: sorted :)

**Review by oliverbutler — APPROVED** [review]
lgtm

File: libs/ts-rest/core/src/lib/dsl.ts [hunk 3]
```diff
@@ -128,9 +131,10 @@ export type AppRouter = {
   [key: string]: AppRouter | AppRoute;
 };
 
-export type RouterOptions = {
+export type RouterOptions<TPrefix extends string> = {
```
Comment (Gabrola) [inline libs/ts-rest/core/src/lib/dsl.ts:134]: I think we can default this to `''` so we don't have to explicitly pass `<string>` everywhere.

```suggestion
export type RouterOptions<TPrefix extends string = ''> = {
```

File: libs/ts-rest/core/src/lib/dsl.spec.ts [hunk 4]
```diff
@@ -477,4 +477,27 @@ describe('contract', () => {
       >
     >;
   });
+
+  describe('pathPrefix', () => {
+    it('Should recursively apply pathPrefix to path', () => {
```
Comment (Gabrola) [inline libs/ts-rest/core/src/lib/dsl.spec.ts:482]: Might be useful to add a type assertion test just like the test above to make sure the runtime path and TS type are the same

**Comment by Gabrola** [comment]
Wow that was quick! Super impressed how you navigated around this one so quickly!

File: libs/ts-rest/core/src/lib/dsl.ts (hunk 3 above)
Comment (Gabrola) [reply libs/ts-rest/core/src/lib/dsl.ts:134]: Actually I might be slightly confused, it's late lol but I think the correct way would be `TPrefix extends string = string`

Comment (Gabrola) [reply libs/ts-rest/core/src/lib/dsl.ts]: @Jwcode-uk I pulled your code locally just to make sure I wasn't just spewing fairy dust, and it indeed worked, so I pushed the changes. I hope you don't mind! 🙏

**Review by Gabrola — APPROVED** [review]

File: libs/ts-rest/core/src/lib/dsl.ts (hunk 3 above)
Comment (Jwcode-uk) [reply libs/ts-rest/core/src/lib/dsl.ts]: Appreciate the help!!

**Pull request merged**